### PR TITLE
Add support for shared resources.

### PR DESCRIPTION
**Pull request by plietar**
This is equivalent to orderly2's orderly_shared_resource.  It looks for files in the `shared` directory, at the root of the repository. Files and directories are copied to the packet directory, and recorded in the metadata.

### CODE REVIEW

**Review by r-ash — APPROVED** [review]
This looks great to me, added @richfitz to see if he wants to have a look too

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -47,6 +47,8 @@ jobs:
           hatch run cov-ci
       - name: Lint
         run: |
+          # https://github.com/python/mypy/issues/10863
+          mkdir -p .mypy_cache
```
Comment (richfitz) [inline .github/workflows/test.yml:51]: > The user may then experience a tremendous feeling of helplessness

oh dear

File: src/orderly/core.py [hunk 2]
```diff
@@ -73,6 +76,70 @@ def resource(files):
...
+        src = shared_path / there
+        dst = packet / here
+
+        if not src.exists():
+            msg = f"Shared resource file '{there}' does not exist. Looked within directory '{shared_path}'"
+            raise Exception(msg)
+        elif src.is_dir():
+            shutil.copytree(src, dst, dirs_exist_ok=True)
+            copied = {
+                os.path.join(here, f): os.path.join(there, f)
```
Comment (richfitz) [inline src/orderly/core.py:136]: Are we not moving to pathlib everywhere? Also happy to sort this out later in a single push too

File: src/outpack/util.py [hunk 3]
```diff
@@ -73,6 +74,12 @@ def assert_file_exists(path, *, workdir=None, name="File"):
         raise Exception(msg)
 
 
+def assert_relative_path(path: str, name: str):
+    if os.path.isabs(path):
```
Comment (richfitz) [inline src/outpack/util.py]: Probably we should exclude `..` from within the path too?

File: src/orderly/core.py [hunk 4]
```diff
@@ -73,6 +76,70 @@ def resource(files):
     return files_expanded
 
 
+def shared_resource(files: Union[str, List[str], Dict[str, str]]):
```
Comment (richfitz) [inline src/orderly/core.py]: nice to support all these forms of files. Do you want to add the return type too?

File: src/orderly/core.py (hunk 2 above)
Comment (plietar) [reply src/orderly/core.py:136]: This is the dictionary that ends up being returned to the caller / used in the packet metadata. I think keeping them as `str` is more appropriate. `Path` isn't json-serializable natively, so we would have to cast it back as str anyway.

File: src/outpack/util.py (hunk 3 above)
Comment (plietar) [reply src/outpack/util.py]: Done.

File: src/orderly/core.py (hunk 4 above)
Comment (plietar) [reply src/orderly/core.py]: Done.

File: src/outpack/util.py [hunk 5]
```diff
@@ -82,6 +82,28 @@ def assert_file_exists(path, *, workdir=None, name="File"):
...
+    # On Windows, this is not equivalent to path.is_absolute().
+    # There are paths which are neither absolute nor relative, but somewhere
+    # between the two. These are paths such as `C:foo` or `\foo`. The former has
+    # a drive but no root, and the latter has a root but no drive. We want to
+    # exclude both scenarios. On POSIX, drive will always be empty and this is
+    # equivalent to calling `is_absolute()`.
+    #
+    # See https://github.com/python/cpython/issues/44626 for some discussion.
+    # Unfortunately, while the issue was closed, the `is_relative` function
+    # mentioned was never added.
```
Comment (richfitz) [inline src/outpack/util.py:97]: Looks like in 3.13+ we can use this! https://github.com/python/cpython/pull/113829 but for now this seems sensible

**Review by richfitz — APPROVED** [review]

**Pull request merged**